### ISSUE TITLE
Make `classNameBindings` a computed property

### ISSUE THREAD

**Issue by YoranBrondsema** [emberjs/rfcs#138]
See the discussion here: https://github.com/emberjs/ember.js/issues/13229.

It would come down to being to do something like this:

```
export default Ember.Component.extend({
  classNameBindings: function() {
    if (conditionA) {
      return ['propertyA', 'propertyB'];
    } else if (conditionB) {
      return ['propertyB'];
    } else {
      return [];
    }
  }.property(...)
});
```

or whatever logic one wishes in `classNameBindings`.

The advantages are that one can decide what properties one adds as a CSS class, potentially saving computations as Ember's lazy evaluation makes sure that a property is never computed if it's never used.

**Comment by stefanpenner** [comment]
The idea of classNamBindings is somewhat legacy and may be superceded with something more idiomatic in the future. The shape of the solution isn't quite known, but the likely-hood of making to many changes to the existing variant at this point is somewhat unlikely. That is unless someone on core feels strongly that an interim step is required.

**Comment by locks** [comment]
@YoranBrondsema a nitpick, you should avoid Function.prototype extensions and use `Ember.computed(…)` instead.

**Comment by YoranBrondsema** [comment]
@stefanpenner OK that makes sense, thanks for clarifying.

@locks I agree, just thought it didn't matter so much in this RFC as the main point is to transfer an idea.

**Comment by nmbl-ui** [comment]
Since angle bracket components are still a ways off, it would be helpful to get some kind of support for computed properties with classNameBindings. Ideally classNameBindings should work just like attributeBindings.

**Comment by lmcardle** [comment]
Since angle bracket components are still a ways off, it would be helpful to get some kind of support for computed properties with classNameBindings. Ideally classNameBindings should work just like attributeBindings.

**Comment by christophermlne** [comment]
@YoranBrondsema, @locks 

I just ran into a situation where I need to bind a class to the result of a computed property. Since there is no way to do this with classNameBindings currently, I had to create an extra wrapping div in my component template to hold the classes. This messed with the flexbox element that contained the component (since the actual component markup is not a direct descendent of the flexbox container), so using flexbox to achieve equal heights for elements in a row is not possible, which led me to move the markup outside of the component, which defeats the purpose of having a component.

What would be nice is an addition to the existing classNameBindings syntax that makes binding to a computed property possible. I'm not crazy about replacing classNameBindings with a computed property, since this would necessitate time-consuming refactoring work for existing Ember applications.

Something like: 
```javascript
   classNameBindings: [':computedClassName']
```

So if there is no left-hand value, assume the right-hand value is a property name.
